armi/interfaces.py

```diff
@@ -77,6 +80,117 @@ class STACK_ORDER:  # pylint: disable=invalid-name, too-few-public-methods
+        Maximum number of iterations allowed before the ``isConverged`` method
+        will return True and the state of the object will be cleared.
+    """
+
+    _SUPPORTED_TYPES = [float, int, list, numpy.ndarray]
+
+    def __init__(self, param, tolerance, maxIters):
+        self.param = param
+        self.tolerance = tolerance
+        self.maxIters = maxIters
+        self._numIters = 0
+        self._previousIterationValue = None
+        self.eps = numpy.inf
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}, Parameter: {self.param}, Convergence Criteria: {self.tolerance}, Maximum Coupled Iterations: {self.maxIters}>"
+
+    def storePreviousIterationValue(self, val):
+        """
+        Stores the previous iteration value of the given parameter.
+
+        Notes
+        -----
+        This checks the type of the value against ``_SUPPORTED_TYPES`` before storing.
+        """
+        if type(val) not in self._SUPPORTED_TYPES:
+            raise TypeError(
+                f"{val} supplied has type {type(val)} which is not supported in {self}. "
+                f"Supported types: {self._SUPPORTED_TYPES}"
+            )
+        self._previousIterationValue = val
+
+    def isConverged(self, val):
+        """
+        Return True if the convergence criteria between the current and previous iteration values are met.
+
+        Notes
+        -----
+        On convergence, this class is automatically reset to its initial condition to avoid retaining
+        or holding a stale state. Calling this method will increment a counter that when exceeded will
+        clear the state. A warning will be reported if the state is cleared prior to the convergence
+        criteria being met.
+
+        Raises
+        ------
+        ValueError
+            If the previous iteration value has not been assigned. The ``storePreviousIterationValue`` method
+            must be called first.
+
+        TypeError
+            If the type of the current value provided is not the same as the previous iteration value.
+        """
+        if self._previousIterationValue is None:
+            raise ValueError(
+                f"Cannot check convergence of {self} with no previous iteration value set. "
+                f"Set using `storePreviousIterationValue` first."
+            )
+
+        if not isinstance(val, type(self._previousIterationValue)):
+            raise TypeError(
+                f"The current value {val} is not the same type as the previous "
+                f"iteration value of {self._previousIterationValue}"
+            )
+
+        previous = self._previousIterationValue
+
+        # Check convergence for integer or float values.
+        if isinstance(val, int) or isinstance(val, float):
+            self.eps = abs(val - previous)
+
+        # Convert list values to a numpy array.
+        if isinstance(val, list):
+            val = numpy.array(val)
+            previous = numpy.array(previous)
```

Comment:
This was addressed somewhat in https://github.com/terrapower/armi/pull/1033/commits/47d10211ae3417caf9ebc8a600be8d82d2241e72. 

I left the call to `storePreviousIterationValue` as separate. We might be able to combine them. I will look into it and update either with a new comment or commit. 

The type check may be redundant, but it might also be nice to have to be safe. There's a good chance that the norm would fail if they were different types anyway (unless they can be type cast into the same type implicitly). So having the TypeError essentially gives us control over the error message.